src/main.rs

```diff
@@ -436,14 +447,62 @@ fn main() -> anyhow::Result<()> {
 
     touch_started_file_indicator();
 
+    general_runtime.block_on(async {
+        wait_stop_signal().await;
+        // We should ignore send error here. It will happen if all the system
+        // components are already finished and dropped rx portion of the notification
+        // channel, which is totally fine in this case.
+        let _ = shutdown_flag_tx.send(true);
+    });
```

Comment:
I don't really like this. I wonder if we can 'wait' for the stop signal and propagate to the `shutdown_flag_tx` without keeping a full extra thread active.

Reply:
Can't we simply remove the `block_on` block, and use `wait_stop_signal` on the main thread?

Reply:
Good point. I did go over this a bit too quickly. Thanks for elaborating on this quickly.